src/LearnMe.Core/Services/Calendar/GoogleCalendar.cs

```diff
@@ -0,0 +1,159 @@
+            EventDateTime end = new EventDateTime()
+            {
+                DateTime = endDateTime,
+                TimeZone = timezone
+            };
+            newCalendarEvent.End = end;
+
+            if (isRecurringEvent)
+            {
+                String[] recurrence = new String[] { "RRULE:FREQ=DAILY;COUNT=7" };
+                newCalendarEvent.Recurrence = recurrence.ToList();
+            }
+
+            // FOR FUTURE - ADD ATTENDEES
+            //EventAttendee[] attendees = new EventAttendee[]
+            //{
+            //    new EventAttendee() { Email = "<---email-address--->@gmail.com" },
+            //};
+            //newEvent.Attendees = attendees.ToList();
+
+            var createdEvent = await _calendarService.Events.Insert(newCalendarEvent, calendarId).ExecuteAsync();
+            newDbEvent.CalendarId = createdEvent.Id;
+
+            return await _repository.InsertAsync(newDbEvent);
+        }
+
+        public async Task<bool> DeleteEventAsync(int id)
+        {
+            return await _repository.DeleteAsync(id);
+        }
+
+        private void SynchronizeDatabaseWithCalendar(string calendarId = "primary")
+        {
+            EventsResource.ListRequest request = _calendarService.Events.List(calendarId);
+            IEnumerable<Event> eventsFromCalendarResult = request.ExecuteAsync().Result.Items;
+
+            var allDatabaseEvents = _repository.GetAllAsync().Result;
+            IList<string> databaseEventsCalendarIds = new List<string>();
+            foreach (var eventFromDatabase in allDatabaseEvents)
+            {
+                databaseEventsCalendarIds.Add(eventFromDatabase.CalendarId);
+            }
+
+            foreach (var eventResult in eventsFromCalendarResult)
+            {
+                // TODO: Add to Calendar specific log
+                //Console.WriteLine($"{eventResult.Summary}\n {eventResult.Location}\n {eventResult.Start.DateTime}");
+
+                if (!databaseEventsCalendarIds.Contains(eventResult.Id))
+                {
+                    // TODO Fix the issue that 1st call of the method goes out from all the methods up to Controller resulting in NullReferenceException
+                    _repository.InsertAsync(new CalendarEvent()
+                    {
+                        Title = eventResult.Summary,
+                        Description = eventResult.Description,
+                        Start = eventResult.Start.DateTime,
+                        End = eventResult.End.DateTime,
+                        IsDone = false,
+                        CalendarId = eventResult.Id
+                    });
+                }
+            }
+        }
+
+        public async Task<IEnumerable<CalendarEventDto>> GetAllEventsAsync(string calendarId = "primary")
+        {
+            // Step 1 - synchronize Google calendar with DB
+            //SynchronizeDatabaseWithCalendar();
+
+            // Step 2 - get all data from DB
```

Comment:
a jak będziesz miała milion rekordów tez je pobierzesz? xd

Reply:
dodałam paginację, na razie nie ma jej w Synchronizacji tylko w kontrolerze.

Synchronizacja to jest pewnie kolejne wyzwanie. Wcześniej muszę raczej poprawić jeszcze tą metodę niżej. Albo może sprawy są powiązane. Bo na razie z tą rekurencją to wychodzi na to, że jak wyników jest więcej niż na 1 stronę to mi zwróci tylko ostatnią (?) czy nie?


public class GoogleCRUD : IGoogleCRUD
    {
        public async Task<IEnumerable<Event>> GetAllEventsAsync(
            CalendarService calendarService,
            bool includeCancelled = false, 
            string calendarId = Constants.CalendarId,
            int maxNumberOfResults = 250, // default value form Google API
            string nextPageToken = null) // corresponds to offset zero (< 250 events in calendar)
        {
            EventsResource.ListRequest request = calendarService.Events.List(calendarId);
            request.ShowDeleted = includeCancelled;
            request.MaxResults = maxNumberOfResults;

            if (nextPageToken != null) request.PageToken = nextPageToken;

            Events result = await request.ExecuteAsync();

            // TODO Update this logic: pagination for retrieving external API data
            if (result.NextPageToken == null)
            {
                return result.Items;
            }
            else
            {
                return await GetAllEventsAsync(
                    calendarService,
                    includeCancelled,
                    calendarId,
                    maxNumberOfResults,
                    result.NextPageToken);
            }
        }
    }

Reply:
Popraw ten komentarz, keep in mind że jeszcze nie odpalałem tej apki tylko patrzyłem na kod więc nie za bardzo rozumiem o jaką rekurencję chodzi. Musisz mi wytłumaczyć step by step co się tutaj dzieje. Ja tylko robię code review, nie zagłębiam się w samo działanie aplikacji :D Nawet nie mam jak bo nie wiem, co ona ma robić